In [48]:
# imports
import os
import requests
import pandas as pd

# Foursquare

Send a request to Foursquare with a small radius (1000m) for all the bike stations in your city of choice. 

Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

Put your parsed results into a DataFrame

In [49]:
fs_api_key = os.environ["foursq_api"]

In [50]:
enlabici_df = pd.read_csv('../data/enlabici-mend.csv')

foursq_results = []

# Loop through each row in the enlabici_df
for index, row in enlabici_df.iterrows():
    latitude = row['latitude']
    longitude = row['longitude']

    # Define the base URL of the Foursquare API
    url = 'https://api.foursquare.com/v3/places/search'

    # Define the parameters (13000 for Dining & Drinking; 10027 for Art and Entertainment > Museum)
    params = {
        'll': f'{latitude},{longitude}',
        'radius': 1000,
        'categories': '13000,10027',
        'fields': 'name,categories,distance,rating,popularity,price',
        'open_at': '1T1931',
        'sort': 'RATING',
        'limit': 20
    }
  
    # Define the headers
    headers = {'Accept': 'application/json'}
    headers['Authorization'] = fs_api_key  

    try:
        response = requests.get(url, params=params, headers=headers)
        response.raise_for_status()
        
    except HTTPError as errh:
        print ("Http Error:",errh)
    except ConnectionError as errc:
        print ("Error Connecting:",errc)
    except Timeout as errt:
        print ("Timeout Error:",errt)
    except TooManyRedirects as errr:
        print ("Too Many Redirects:",errr)
    except RequestException as err:
        print ("Something Else:",err)
    else:
        data = response.json()
        
    places = data.get('results')
    for place in places:
        place['latitude'] = latitude
        place['longitude'] = longitude
        foursq_results.append(place)


In [51]:
fs_df = pd.DataFrame(foursq_results)
fs_df

categories  distance  \
0    [{'id': 13070, 'name': 'Argentinian Restaurant...       523   
1    [{'id': 13002, 'name': 'Bakery', 'short_name':...       895   
2    [{'id': 13035, 'name': 'Coffee Shop', 'short_n...       514   
3    [{'id': 13070, 'name': 'Argentinian Restaurant...       865   
4    [{'id': 13070, 'name': 'Argentinian Restaurant...       194   
..                                                 ...       ...   
315  [{'id': 13065, 'name': 'Restaurant', 'short_na...       797   
316  [{'id': 13064, 'name': 'Pizzeria', 'short_name...       822   
317  [{'id': 13065, 'name': 'Restaurant', 'short_na...       850   
318  [{'id': 13002, 'name': 'Bakery', 'short_name':...       900   
319  [{'id': 10000, 'name': 'Arts and Entertainment...       927   

                                      name  popularity  price  rating  \
0                                  Azafran    0.998870    3.0     8.9   
1                           BRÖD Panadería    0.991713    1.0     8.9   
2                           Famiglia Perin    0.987067    1.0     8.8   
3                        Ceibo Restaurante    0.944375    2.0     8.6   
4                              Anna Bistro    0.992843    3.0     8.5   
..                                     ...         ...    ...     ...   
315                   Dorrego Sur Sandwich    0.501128    NaN     NaN   
316                          Gallego Pizza    0.956247    1.0     NaN   
317                        Cardamomo Resto    0.248083    NaN     NaN   
318                          San Francisco    0.775113    1.0     NaN   
319  Principe de Asturias-Eventos Catering    0.645028    1.0     NaN   

      latitude  longitude  
0   -32.884090 -68.848970  
1   -32.884090 -68.848970  
2   -32.884090 -68.848970  
3   -32.884090 -68.848970  
4   -32.884090 -68.848970  
..         ...        ...  
315 -32.924682 -68.842504  
316 -32.924682 -68.842504  
317 -32.924682 -68.842504  
318 -32.870709 -68.858351  
319 -32.870709 -68.858351  

[320 rows x 8 columns]

In [52]:
fs_df.to_csv('../data/foursquare_df_orig.csv', index=False)

In [53]:
# Check missing coordinate -32.9434,-68.818481
url = 'https://api.foursquare.com/v3/places/search'

params = {
    'll': '-32.9434,-68.818481',
    'radius': 1000,
    'categories': '13000,10027',
    'fields': 'name,categories,distance,rating,popularity,price',
    'open_at': '1T1931',
    'sort': 'RATING',
    'limit': 20
}
headers = {'Accept': 'application/json'}
headers['Authorization'] = fs_api_key  

response = requests.get(url, params=params, headers=headers)

data = response.json()
add_place = data.get('results')

In [54]:
print(response)
print(data)
print(add_place)

<Response [200]>
{'results': [], 'context': {'geo_bounds': {'circle': {'center': {'latitude': -32.9434, 'longitude': -68.818481}, 'radius': 1000}}}}
[]


# Yelp

Send a request to Yelp with a small radius (1000m) for all the bike stations in your city of choice.

Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

Put your parsed results into a DataFrame

In [55]:
yelp_api_key = os.environ["yelp_api"]

In [56]:
enlabici_df = pd.read_csv('../data/enlabici-mend.csv')

yelp_results = []

# Loop through each row in the enlabici_df
for index, row in enlabici_df.iterrows():
    latitude = row['latitude']
    longitude = row['longitude']

    # Define the base URL of the Foursquare API
    url = 'https://api.yelp.com/v3/businesses/search'
    
    # Define the parameters 
    params = {
        'latitude': latitude,
        'longitude': longitude,
        'term': 'restaurants',
        'radius': 1000,
        'categories': 'restaurants,All',
        'sort_by': 'rating',
        'limit': 20
    }
    
    # Define the headers
    headers = {
        'accept': 'application/json',
        'Authorization': f'Bearer {yelp_api_key}'
    }
    
    try:
        response = requests.get(url, params=params, headers=headers)
        response.raise_for_status()
        
    except HTTPError as errh:
        print ("Http Error:",errh)
    except ConnectionError as errc:
        print ("Error Connecting:",errc)
    except Timeout as errt:
        print ("Timeout Error:",errt)
    except TooManyRedirects as errr:
        print ("Too Many Redirects:",errr)
    except RequestException as err:
        print ("Something Else:",err)
    else:
        data = response.json()
        
    places = data.get('businesses')
    for place in places:
        place['latitude'] = latitude
        place['longitude'] = longitude
        yelp_results.append(place)

In [57]:
yp_df = pd.DataFrame(yelp_results)
yp_df

id  \
0    HtAhkhT4ITrDFSQxqbbucQ   
1    I3yljXkYDFRAZ3TuObhC5w   
2    V2sDzIU3rKFkA7r_8-POuQ   
3    p_nchEuqalCPSzYjR2QCJg   
4    w4Mt3rz9-pcSKMYcnAeqrg   
..                      ...   
273  EuVG8YRvW9zDGb5hx7SOpg   
274  1cZzCYzqkGGmmRwDmLcyXQ   
275  lrun9tbFJxnr3YdqDe4k4A   
276  jhjx3L8t-YwP23cUjvZ5vA   
277  cPUIV8YS2ul2w7IWiWr6Yg   

                                                 alias  \
0                                     cachitas-mendoza   
1                                     la-barra-mendoza   
2                                        unión-mendoza   
3                  cordillera-vinos-y-fuegos-mendoza-2   
4                                     silla-14-mendoza   
..                                                 ...   
273             gula-gula-la-pizza-delivery-godoy-cruz   
274  aldo-pizzeria-sandwicheria-restaurante-deliver...   
275               la-paulina-pizzeria-delivery-mendoza   
276                pizzas-artesanales-amarello-mendoza   
277  aldo-restaurante-y-pizzeria-sandwicheria-deliv...   

                                                  name  \
0                                            Cachita's   
1                                             La Barra   
2                                                Unión   
3                            Cordillera Vinos y Fuegos   
4                                             Silla 14   
..                                                 ...   
273                    Gula Gula - la Pizza - Delivery   
274  Aldo Pizzeria Sandwicheria Restaurante - Delivery   
275                     la Paulina Pizzeria - Delivery   
276                        Pizzas Artesanales Amarello   
277  Aldo Restaurante y Pizzeria Sandwicheria - Del...   

                                             image_url  is_closed  \
0    https://s3-media1.fl.yelpcdn.com/bphoto/cli340...      False   
1    https://s3-media3.fl.yelpcdn.com/bphoto/gVwVZO...      False   
2    https://s3-media2.fl.yelpcdn.com/bphoto/qkh9dz...      False   
3    https://s3-media1.fl.yelpcdn.com/bphoto/5zuotZ...      False   
4    https://s3-media2.fl.yelpcdn.com/bphoto/hkIT9V...      False   
..                                                 ...        ...   
273                                                         False   
274                                                         False   
275                                                         False   
276                                                         False   
277                                                         False   

                                                   url  review_count  \
0    https://www.yelp.com/biz/cachitas-mendoza?adju...             7   
1    https://www.yelp.com/biz/la-barra-mendoza?adju...             5   
2    https://www.yelp.com/biz/uni%C3%B3n-mendoza?ad...             5   
3    https://www.yelp.com/biz/cordillera-vinos-y-fu...            13   
4    https://www.yelp.com/biz/silla-14-mendoza?adju...             8   
..                                                 ...           ...   
273  https://www.yelp.com/biz/gula-gula-la-pizza-de...             0   
274  https://www.yelp.com/biz/aldo-pizzeria-sandwic...             1   
275  https://www.yelp.com/biz/la-paulina-pizzeria-d...             0   
276  https://www.yelp.com/biz/pizzas-artesanales-am...             0   
277  https://www.yelp.com/biz/aldo-restaurante-y-pi...             0   

                                            categories  rating  \
0    [{'alias': 'cocktailbars', 'title': 'Cocktail ...     5.0   
1       [{'alias': 'argentine', 'title': 'Argentine'}]     5.0   
2       [{'alias': 'cafeteria', 'title': 'Cafeteria'}]     5.0   
3    [{'alias': 'wine_bars', 'title': 'Wine Bars'},...     4.5   
4       [{'alias': 'cafeteria', 'title': 'Cafeteria'}]     4.5   
..                                                 ...     ...   
273             [{'alias': 'pizza', 'title': 'Pizza'}]     0.0   
274             [{'alias': 'pizza', 

In [58]:
yp_df.to_csv('../data/yelp_df_orig.csv', index=False)

# Comparing Results

**Which API provided you with more complete data? Provide an explanation.**

Overall, the Foursquare API seems better in terms of completeness of data, ease of use, and ability to select data
* With a limit of 20 for each latitude/longitude row, Foursquare pulled 320 total results while Yelp pulled 278. 
* Looking at the column with the most nulls, Foursquare has a non-null value for 73% of the entries (ratings 233 out of 320). Yelp has a value for 50% of the entries (price 140 out of 278)
* Foursquare rating column ranges from 0-10, while Yelp ranges from 0-5. When sorting by rating, Foursquare results show more distinction between entries.
* Foursquare lets you finetune the fields returned a bit more easily than Yelp

Yelp provides more data columns but completeness and the ability to finetune is more useful rather than getting lots of data for the sake of it. However, a point in favour of Yelp is that it covered all 22 bike stations. Foursquare returned results for 21 latitude/longitude coordinates. Submitting a request for the missing coordinate returned no response. 

In [59]:
fs_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 320 entries, 0 to 319
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   categories  320 non-null    object 
 1   distance    320 non-null    int64  
 2   name        320 non-null    object 
 3   popularity  320 non-null    float64
 4   price       264 non-null    float64
 5   rating      233 non-null    float64
 6   latitude    320 non-null    float64
 7   longitude   320 non-null    float64
dtypes: float64(5), int64(1), object(2)
memory usage: 20.1+ KB


In [60]:
yp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278 entries, 0 to 277
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             278 non-null    object 
 1   alias          278 non-null    object 
 2   name           278 non-null    object 
 3   image_url      278 non-null    object 
 4   is_closed      278 non-null    bool   
 5   url            278 non-null    object 
 6   review_count   278 non-null    int64  
 7   categories     278 non-null    object 
 8   rating         278 non-null    float64
 9   coordinates    278 non-null    object 
 10  transactions   278 non-null    object 
 11  price          140 non-null    object 
 12  location       278 non-null    object 
 13  phone          278 non-null    object 
 14  display_phone  278 non-null    object 
 15  distance       278 non-null    float64
 16  latitude       278 non-null    float64
 17  longitude      278 non-null    float64
dtypes: bool(1)

**Get the top 10 restaurants according to their rating**

In [61]:
# FOURSQUARE TOP 10 

# create a copy of fs_df
fs_df1 = fs_df.copy()

# drop duplicates between the columns: name, rating, popularity, price; keep first instance
fs_df1_unique = fs_df1.drop_duplicates(subset=['name','rating','popularity','price'], keep='first')

# show top 10 rows ordered by 'rating' column
fs_top_10 = fs_df1_unique.nlargest(10, 'rating')

print(fs_top_10[['name','rating','popularity']])

                             name  rating  popularity
0                         Azafran     8.9    0.998870
1                  BRÖD Panadería     8.9    0.991713
2                  Famiglia Perin     8.8    0.987067
3               Ceibo Restaurante     8.6    0.944375
4                     Anna Bistro     8.5    0.992843
94                   Bodegas CARO     8.5    0.956429
121                     Calendula     8.4    0.956806
25                      Starbucks     8.3    0.989076
44              Cuarto Intermedio     8.2    0.804746
5    Brillat Savarin - Pastelería     8.1    0.945756


In [62]:
# YELP TOP 10

# create a copy of yp_df
yp_df1 = yp_df.copy()

# drop duplicates between the columns: name, rating, review_count; keep first instance
yp_df1_unique = yp_df1.drop_duplicates(subset=['name', 'rating','review_count'], keep='first')

# show top 10 rows ordered by 'rating' column
yp_top_10 = yp_df1_unique.nlargest(10, 'rating')

print(yp_top_10[['name', 'rating', 'review_count']])

                  name  rating  review_count
0            Cachita's     5.0             7
1             La Barra     5.0             5
2                Unión     5.0             5
6    Tasting Patagonia     5.0             4
11        Caché Bistro     5.0             2
12     Santa y Pecador     5.0             2
51  Burger Social Club     5.0             1
52      Café Plazoleta     5.0             1
53   Tasca de La Plaza     5.0             1
61   Cantina la Rambla     5.0             1


The above results also highlight where Yelp falls short. The top 10 Yelp places based on ratings don't actually have many reviews. They could just be highly rated (for now) because they're fairly new.

For example, "Anna Bistro" falls under the top 10 in the Foursquare results, however, for Yelp it doesn't make the cut because it has a 4.5 rating BUT 45 reviews. All top 10 results for Yelp have under 10 reviews.